### PR TITLE
NIAD-1976: Map Medication Requests

### PR DESCRIPTION
**Pull request by ElliotB-96**
Initial implementation of medicationRequestMapper

### CODE REVIEW

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java [hunk 1]
```diff
@@ -91,17 +94,20 @@ public Bundle mapToBundle(RCMRIN030000UK06Message xmlMessage) {
...
 
-        var observations = mapObservations(ehrExtract, patient, encounters);
+        var medicationResources = medicationRequestMapper.mapResources(ehrExtract, encounters, patient);
+        addEntries(bundle, medicationResources);
+
+        var observations = mapObservations(ehrExtract, patient, List.of()); //TODO: Provide list of encounters
         addEntries(bundle, observations);
 
-        var immunizations = mapImmunizations(ehrExtract, patient, encounters);
+        var immunizations = mapImmunizations(ehrExtract, patient, List.of()); //TODO: Provide list of encounters (NIAD-1961)
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java:103]: It seems like Rory's changes in this file were discarded :(

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    public List<DomainResource> mapResources(RCMRMT030101UK04EhrExtract ehrExtract, List<Encounter> encounters, Patient patient) {
+
+        List<RCMRMT030101UK04EhrComposition> ehrCompositions = ehrExtract.getComponent()
+            .stream()
+            .map(RCMRMT030101UK04Component::getEhrFolder)
+            .map(RCMRMT030101UK04EhrFolder::getComponent)
+            .flatMap(List::stream)
+            .map(RCMRMT030101UK04Component3::getEhrComposition)
+            .collect(Collectors.toList());
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:52]: Could be replaced with .toList(); (it's the case for many places in this class, please replace them all :D)

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    private MedicationRequest mapToOrderMedicationRequest(RCMRMT030101UK04MedicationStatement medicationStatement,
+        RCMRMT030101UK04Prescribe supplyPrescribe) {
+        return medicationRequestOrderMapper.mapToOrderMedicationRequest(medicationStatement, supplyPrescribe);
+    }
+
+    private MedicationRequest mapToPlanMedicationRequest(RCMRMT030101UK04EhrExtract ehrExtract,
+        RCMRMT030101UK04MedicationStatement medicationStatement, RCMRMT030101UK04Authorise supplyAuthorise) {
+        return medicationRequestPlanMapper.mapToPlanMedicationRequest(ehrExtract, medicationStatement, supplyAuthorise);
+    }
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:137]: I'm not sure if those methods are needed - they just simply pass the parameters without changing them, to methods with the same name

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            DateFormatUtil.parseToDateTimeType(ehrExtract.getAvailabilityTime().getValue()));
+        var requester = extractRequester(ehrComposition, medicationStatement);
+        var recorder = extractRecorder(ehrComposition, medicationStatement);
+
+        List<Medication> medications = medicationStatement.getConsumable()
+            .stream()
+            .map(medicationMapper::createMedication)
+            .collect(Collectors.toList());
+
+        List<MedicationRequest> medicationRequestsOrder = medicationStatement.getComponent()
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:83]: I'd consider creating separate methods for all of those streams here that return List<MedicationRequest>

**Review by mrogowska — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java (hunk 2 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:52]: Added, except for line 121 it seemed to need it :D

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java (hunk 1 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java:103]: Added the encounter list to each method, will have to update the INT test file to accomodate this

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java (hunk 3 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:137]: removed these extra methods

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/CodeableConceptMapper.java [hunk 5]
```diff
@@ -23,7 +23,7 @@ public CodeableConcept mapToCodeableConcept(CD codedData) {
     }
 
     private boolean hasSnomedCodeInMainCode(CD codedData) {
-        return codedData.getCodeSystem().equals(SNOMED_SYSTEM_CODE);
+        return codedData.hasCodeSystem() && codedData.getCodeSystem().equals(SNOMED_SYSTEM_CODE);
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/CodeableConceptMapper.java:26]: equate against the SNOMED_SYSTEM_CODE please

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    private IdGeneratorService idGeneratorService;
+
+    protected Medication createMedication(RCMRMT030101UK04Consumable consumable) {
+        if (hasManufacturedMaterial(consumable)) {
+            CD code = consumable.getManufacturedProduct().getManufacturedMaterial().getCode();
+
+            Medication medication = new Medication();
+            medication.setId(getMedicationId(code));
+            medication.setMeta(generateMeta(MEDICATION_URL));
+            medication.setCode(codeableConceptMapper.mapToCodeableConcept(code));
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java:44]: can chain these (if you find this more readable then can leave it)

Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java:44]: Apologies, but could use the util you've made for this and add it to all the mappers that use it?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        }
+    }
+
+    private Optional<Reference> extractRequester(RCMRMT030101UK04EhrComposition ehrComposition,
+        RCMRMT030101UK04MedicationStatement medicationStatement) {
+        if (medicationStatement.hasParticipant()) {
+            var pprfRequester = medicationStatement.getParticipant()
+                .stream()
+                .filter(participant -> !participant.hasNullFlavour())
+                .filter(participant -> participant.getTypeCode().contains("PPRF") || participant.getTypeCode().contains("PRF"))
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:153]: Could you add constants for these?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestOrderMapper.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        RCMRMT030101UK04Prescribe supplyPrescribe) {
+        var ehrSupplyPrescribeIdExtract = extractEhrSupplyPrescribeId(supplyPrescribe);
+        if (ehrSupplyPrescribeIdExtract.isPresent()) {
+            var ehrSupplyPrescribeId = ehrSupplyPrescribeIdExtract.get();
+            var supplyAuthorise = extractSupplyAuthorise(medicationStatement, ehrSupplyPrescribeId);
+            MedicationRequest medicationRequest = createMedicationRequestSkeleton(ehrSupplyPrescribeId);
+
+            medicationRequest.addIdentifier(buildIdentifier(ehrSupplyPrescribeId, ""));
+            medicationRequest.setStatus(MedicationRequest.MedicationRequestStatus.COMPLETED);
+            medicationRequest.setIntent(MedicationRequest.MedicationRequestIntent.ORDER);
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestOrderMapper.java:48]: Could you specfically import these Status and Intents?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationStatementMapper.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        return null;
+    }
+
+    private MedicationStatement.MedicationStatementStatus buildMedicationStatementStatus(RCMRMT030101UK04Authorise supplyAuthorise) {
+        if (supplyAuthorise.hasStatusCode() && supplyAuthorise.getStatusCode().hasCode()
+            && COMPLETE.equals(supplyAuthorise.getStatusCode().getCode())) {
+            return MedicationStatement.MedicationStatementStatus.COMPLETED;
+        } else {
+            return MedicationStatement.MedicationStatementStatus.ACTIVE;
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationStatementMapper.java:78]: Specifically import these pls

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import lombok.SneakyThrows;
+import uk.nhs.adaptors.pss.translator.mapper.CodeableConceptMapper;
+import uk.nhs.adaptors.pss.translator.service.IdGeneratorService;
+
+@ExtendWith(MockitoExtension.class)
+public class MedicationMapperTest {
+
+    private static final String XML_RESOURCES_BASE = "xml/Consumable/";
+    private static final String TEST_ID = "TEST_ID";
+    @Mock
```
Comment (rorysproule1) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperTest.java:29]: New line above this pls (sorry)

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestOrderMapperTest.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        assertThat(medicationRequest.getNote().size()).isEqualTo(1);
+        assertThat(medicationRequest.getDosageInstructionFirstRep().getText()).isEqualTo(TAKE_ONE_DAILY);
+        assertThat(medicationRequest.getDispenseRequest().getQuantity().getValue()).isNull();
+        assertThat(medicationRequest.getDispenseRequest().getValidityPeriod().getStartElement().getValue())
+            .isEqualTo(DateFormatUtil.parseToDateTimeType(AVAILABILITY_TIME).getValue());
+    }
+
+    public void assertCommonValues(MedicationRequest medicationRequest) {
+        assertThat(medicationRequest.getIdentifier().size()).isEqualTo(1);
+        assertThat(medicationRequest.getIntent()).isEqualTo(MedicationRequest.MedicationRequestIntent.ORDER);
```
Comment (rorysproule1) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestOrderMapperTest.java:101]: Specifically import these again pls

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestPlanMapperTest.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        assertStatusReasonInformation(statusReason.get(0));
+
+        var prescriptionType = medicationRequest
+            .getExtensionsByUrl("https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-PrescriptionType-1");
+        assertThat(prescriptionType.size()).isEqualTo(ONE);
+        var codeableConcept = (CodeableConcept) prescriptionType.get(0).getValue();
+        assertThat(codeableConcept.getCodingFirstRep().getDisplay()).isEqualTo("Repeat");
+
+        assertThat(medicationRequest.getStatus()).isEqualTo(MedicationRequest.MedicationRequestStatus.ACTIVE);
+        assertThat(medicationRequest.getIntent()).isEqualTo(MedicationRequest.MedicationRequestIntent.PLAN);
```
Comment (rorysproule1) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestPlanMapperTest.java:91]: Specifically import these aswell pls

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationStatementMapperTest.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        var dateTime = (DateTimeType) lastIssuedDate.get(0).getValue();
+        assertThat(dateTime.getValue()).isEqualTo(DateFormatUtil.parseToDateTimeType("20060428").getValue());
+
+        var prescribingAgency = medicationStatement1
+            .getExtensionsByUrl("https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-PrescribingAgency-1");
+        assertThat(prescribingAgency.size()).isEqualTo(1);
+        assertThat(medicationStatement1.getBasedOnFirstRep().getReferenceElement().getIdPart()).isEqualTo(TEST_ID);
+        assertThat(medicationStatement1.getStatus()).isEqualTo(MedicationStatement.MedicationStatementStatus.ACTIVE);
+        assertThat(medicationStatement1.getMedicationReference().getReferenceElement().getIdPart()).isEqualTo(MEDICATION_ID);
+        assertThat(medicationStatement1.getTaken()).isEqualTo(MedicationStatement.MedicationStatementTaken.UNK);
```
Comment (rorysproule1) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationStatementMapperTest.java:68]: Search and replace any of these ill stop commenting about them apologies

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+            .stream()
+            .map(RCMRMT030101UK04Component::getEhrFolder)
+            .map(RCMRMT030101UK04EhrFolder::getComponent)
+            .flatMap(List::stream)
+            .map(RCMRMT030101UK04Component3::getEhrComposition)
+            .collect(Collectors.toList());
+
+        for (RCMRMT030101UK04EhrComposition ehrComposition : ehrCompositions) {
+            var context = encounters.stream()
+                .filter(encounter1 -> encounter1.getId().equals(ehrComposition.getId().getRoot())).findFirst();
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:56]: Is there a method/util that finds a matching encounter from the ehrCOmpositions? I think there should be by now, maybe use it if there is one or create it for other mappers to use?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java (hunk 2 above)
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:51]: I was not aware of this method, thank you very much :D

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java (hunk 6 above)
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java:44]: Will leave as is, .setCode is not a method in Resource.class which is the return type once we start chaining methods

Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java:44]: have changed the other mappers to use meta generation however have left identifiers as they all seem to do it slightly differently.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java (hunk 14 above)
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:55]: Only util method I could find returns a reference, this approach requires the full encounter resource.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java (hunk 6 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java:43]: Think theres a ticket for the identifier source id stuff anyway so will likely be covered in that. Cheers!

**Review by rorysproule1 — APPROVED** [review]

**Review by mrogowska — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+            return medication;
+        }
+        return null;
+    }
+
+    public String getMedicationId(CD code) {
+        var key = keyBuilder(CD::hasCode, CD::getCode, code)
```
[the remark this answers is not in the document]
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapper.java:51]: will remove the hyphen in general, doesn't overly serve a purpose

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+        }
+    }
+
+    private Optional<Reference> extractRequester(RCMRMT030101UK04EhrComposition ehrComposition,
+        RCMRMT030101UK04MedicationStatement medicationStatement) {
+        if (medicationStatement.hasParticipant()) {
+            var pprfRequester = medicationStatement.getParticipant()
+                .stream()
+                .filter(participant -> !participant.hasNullFlavour())
+                .filter(participant -> participant.getTypeCode().contains(PPRF) || participant.getTypeCode().contains(PRF))
```
[the remark this answers is not in the document]
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java:171]: typeCode is marked as required it also returns an empty list if null public List<String> getTypeCode() {
        if (typeCode == null) {
            typeCode = new ArrayList<String>();
        }
        return this.typeCode;
    }

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestPlanMapperTest.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+        when(medicationMapper.extractMedicationReference(any()))
+            .thenReturn(Optional.of(new Reference(new IdType(ResourceType.Medication.name(), MEDICATION_ID))));
+
+        var medicationRequest
+            = medicationRequestPlanMapper.mapToPlanMedicationRequest(ehrExtract, medicationStatement.get(), supplyAuthorise.get());
+
+        var repeatInformation = medicationRequest
+            .getExtensionsByUrl("https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationRepeatInformation-1");
+        assertThat(repeatInformation.size()).isEqualTo(ONE);
```
[the remark this answers is not in the document]
Comment (ElliotB-96) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestPlanMapperTest.java:77]: I think I had one fail on checkstyle for magic numbers so I decided to swap all of them in this test, must have missed the 2, updated now

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperUtils.java [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+        try {
+            SimpleQuantity quantity = new SimpleQuantity();
+            quantity.setValue(Long.parseLong(quantitySupplied.getValue()));
+            if (quantitySupplied.hasTranslation()
+                && quantitySupplied.getTranslation().get(0).hasOriginalText()) {
+                quantity.setUnit(quantitySupplied.getTranslation().get(0).getOriginalText());
+            }
+            return Optional.of(quantity);
+        } catch (NumberFormatException nfe) {
+            LOGGER.info("Unable to parse value from quantity supplied: {}", quantitySupplied.getValue());
```
[the remark this answers is not in the document]
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperUtils.java:98]: @dawidskainos what do you think the best approach is for failing stuff like this? fail the individual resource or the whole transaction

[the remark this answers is not in the document]
Comment (ElliotB-96) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationMapperUtils.java:98]: Discussed with @dawidskainos and @mrogowska will remove catch.

**Review by mrogowska — APPROVED** [review]

**Review by dawidskainos — APPROVED** [review]

**Pull request merged**